craft_providers/bases/buildd.py

```diff
@@ -0,0 +1,522 @@
+    )
+
+
+def _check_deadline(
+    deadline: Optional[float],
+    *,
+    message: str = "Timed out configuring environment.",
+) -> None:
+    """Check deadline and raise error if passed.
+
+    :param deadline: Optional time.time() deadline.
+
+    :raises BaseConfigurationError: if deadline is passed.
+    """
+    if deadline is not None and time.time() >= deadline:
+        raise BaseConfigurationError(brief=message)
+
+
+class BuilddBaseAlias(enum.Enum):
+    """Mappings for supported buildd images."""
+
+    XENIAL = "16.04"
+    BIONIC = "18.04"
+    FOCAL = "20.04"
+
+
+class BuilddBase(Base):
+    """Support for Ubuntu minimal buildd images.
+
+    :param alias: Base alias / version.
+    :param hostname: Hostname to configure.
+    :param command_environment: Additional environment to configure for command.
+        If specifying an environment, default_command_environment() is provided
+        for the minimum required environment configuration.
+    """
+
+    def __init__(
+        self,
+        *,
+        alias: BuilddBaseAlias,
+        hostname: str = "craft-buildd-instance",
+        command_environment: Optional[Dict[str, Optional[str]]] = None,
+    ):
+        self.alias: BuilddBaseAlias = alias
+        self.hostname: str = hostname
+
+        if command_environment is not None:
+            self.command_environment = command_environment
+        else:
+            self.command_environment = default_command_environment()
+
+    def setup(
+        self,
+        *,
+        executor: Executor,
+        retry_wait: float = 0.25,
+        timeout: Optional[float] = None,
+    ) -> None:
+        """Prepare base instance for use by the application.
+
+        Wait for environment to become ready and configure it.  At completion of
+        setup, the executor environment should have networking up and have all
+        of the installed dependencies required for subsequent use by the
+        application.
+
+        Setup may be called more than once in a given instance to refresh/update
+        the environment.
+
+        If timeout is specified, abort operation if time has been exceeded.
+
+        Guarantees provided by this setup:
+
+            - configured hostname
+
+            - networking available (IP & DNS resolution)
+
+            - apt cache up-to-date
+
+            - snapd configured and ready
+
+            - system services are started and ready
+
+        :param executor: Executor for target container.
+        :param retry_wait: Duration to sleep() between status checks (if
+            required).
+        :param timeout: Timeout in seconds.
+
+        :raises BaseConfigurationError: on other unexpected error.
+        """
+        if timeout is not None:
+            deadline: Optional[float] = time.time() + timeout
+        else:
+            deadline = None
+
+        self._setup_environment(executor=executor, deadline=deadline)
```

Comment:
This probably won't make much difference in a real world usage, but it seems that even if the deadline has already been reached it will still execute the first action in `_setup_environment()`. If I start with an expired deadline I would expect it to do nothing at all.

Reply:
Agreed, fixed.